Add nerve section morphometrics + axon density

hermancollin opened the pull request:
## Checklist
- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://intranet.neuro.polymtl.ca/geek-tips/contributing#pr-labels-a-href-pr-labels-id-pr-labels-a) to this PR
- [x] I've added relevant tests for my contribution
- [x] I've updated the documentation and/or added correct docstrings
- [x] I've assigned a reviewer
- [x] I've consulted [ADS's internal developer documentation](https://github.com/neuropoly/axondeepseg/wiki/Guidelines-for-contribution) to ensure my contribution is in line with any relevant design decisions


## Description
This PR adds a feature to compute the nerve section areas. It assumes the user provided a `_seg-nerve` file alongside the image and can be used with the `-n` option in the morphometrics script.

## Linked issues
Related to #866

hermancollin commented:
Instead of a csv file it might be better to print the results in a JSON like so:

```python
{
   fascicle_areas: {
      0: 43432,
      1: 4313,
      2: 56753
   },
   total_area: 999999,
   total_nb_axons_inside_area: 4183,
   axon_density: 0.444
}
```

edgark31 commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -209,13 +274,24 @@ def main(argv=None):
+                        pred_myelin = load_mask(current_path_target, 'myelin', myelin_suffix)
+                    except:
+                        pred_axon = None
+                        pred_myelin = None
```
No particular reason. I will remove it.

hermancollin commented:
Thank you @edgark31 this looks good.

@mathieuboudreau maybe you would like to take a look at this PR before we start adding tests? The changes include:
- I refactored `launch_morphometrics_computation` such that it now chooses a mode, either `myelinated`, `unmyelinated` or `nerve`. Depending on the mode, some filenames and inputs are set differently.
- `-n` option to compute nerve areas and axon density. Assuming a `_seg-nerve.png` mask is present, it will compute the total area of the nerves, remove axons outside the nerve mask and compute the axon density. To test the feature, I suggest to use the images with nerve masks in the AxonDeepSeg gdrive located in `results/20250130_columbia_segmentations/final_results_JU33_MN90_RT78/JU33`.

The output of nerve mode is a JSON like the following:
```json
{
    "fascicle_areas": {
        "0": {
            "value": 63530.65192457,
            "unit": "um^2"
        },
        "1": {
            "value": 27369.55348444,
            "unit": "um^2"
        },
        "2": {
            "value": 7756.29270158,
            "unit": "um^2"
        },
        "3": {
            "value": 2921.02995175,
            "unit": "um^2"
        }
    },
    "total_area": {
        "value": 101577.52806233999,
        "unit": "um^2"
    },
    "axon_density": {
        "value": 0.059314300268287164,
        "unit": "axon/um^2"
    }
}
```

hermancollin commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -148,12 +149,44 @@ def remove_outside_nerve(pred_axon, pred_myelin, pred_nerve):
+    # get fascicle centroids 
+    fascicle_areas = {idx: fascicle["value"] for idx, fascicle in nerve_data["fascicle_areas"].items()}
+    
+    # extract (x, y) coordinates of axons
+    axon_positions = axon_df[["x0 (px)", "y0 (px)"]].values
+
+    # use KMeans clustering to assign axons to fascicles
+    kmeans = KMeans(n_clusters=len(fascicle_areas), random_state=42, n_init=10)
+    axon_df["fascicle_id"] = kmeans.fit_predict(axon_positions)
```
ok first of all, this was a very nice idea to solve this problem. However, I think this might introduce some complexity that we don't want to deal with. I know k-means is pretty reliable, but there are some edge cases where this might fail.

There is a simpler way to do this. When you call `get_axon_morphometrics()`, internally there is a watershed segmentation that is performed to separate all objects. This is how morphometrics are computed for every single object in a mask. Without going too much into details, you should use the same function that is used here: https://github.com/axondeepseg/axondeepseg/blob/9e92a0c114e19b830bcd4e10948e7a53e81b35a4/AxonDeepSeg/morphometrics/compute_morphometrics.py#L120

hermancollin replied:
(axon_objects contains the actual regions of every individual object, separated)

hermancollin replied:
based on you can filter axons like you did previously

hermancollin reviewed (commented):
As a sidenote, I think all the functions that you implemented in this file should be moved to `AxonDeepSeg.morphometrics.compute_morphometrics` which might be more appropriate for these.

hermancollin commented on AxonDeepSeg/morphometrics/compute_morphometrics.py:
```diff
@@ -589,37 +589,51 @@ def remove_outside_nerve(pred_axon, pred_myelin, pred_nerve):
+        if 0 <= y_int < height and 0 <= x_int < width:
+            label_id = nerve_fascicles[y_int, x_int]
+            if label_id > 0:
+                axon_df.loc[i, "fascicle_id"] = label_id
```
hm I think that's where the problem is. `skimage.measure` considers 0 as background, and 1 is the first object.

```suggestion
                axon_df.loc[i, "fascicle_id"] = label_id - 1
```

edgark31 replied:
I think you're right, skimage.measure.label starts at 1 for objects, so subtracting 1 here makes sense to align with zero-based indexing elsewhere. I'll test it to make sure it works.

hermancollin reviewed (commented):
Left some comments again. Overall, this is great. Apart from the nerve indexing issue and the `cleanup_nerve_mask` function, I'm pretty satisfied with everything here. 

One small nitpick is that the `unit` fields in the JSON make the file less readable. We should remove these and document it differently, i.e. in the doc itself.

edgark31 commented on AxonDeepSeg/morphometrics/compute_morphometrics.py:
```diff
@@ -555,13 +555,13 @@ def save_nerve_morphometrics_to_json(stats_dataframe, output_path):
 
     fascicle_areas = {
-        str(idx): {"value": row["nerve_area"], "unit": "um^2"}
+        str(idx + 1): {"value": row["nerve_area"], "unit": "um^2"}
```
the +1 ensures that the JSON labels align with the skimage labeling convention (where 0 is background and actual regions start from 1)

hermancollin replied:
@edgark31 we need to remove the +1 here as well

```suggestion
        str(idx): {"value": row["nerve_area"], "unit": "um^2"}
```

edgark31 commented on AxonDeepSeg/morphometrics/compute_morphometrics.py:
```diff
@@ -589,37 +589,51 @@ def remove_outside_nerve(pred_axon, pred_myelin, pred_nerve):
 
     return pred_axon, pred_myelin
 
+def cleanup_nerve_mask(nerve_mask, min_object_size=5000, max_hole_size=5000):
+    """
+    Removes small objects, fills holes up to max_hole_size, etc.
+    nerve_mask: a binary numpy array where True=inside nerve, False=background
+    """
+    filled = morphology.remove_small_holes(nerve_mask, area_threshold=max_hole_size)
+    cleaned = morphology.remove_small_objects(filled, min_size=min_object_size)
+    
+    return cleaned
```
yes, the function is meant to clean up segmentation artifacts in the seg-nerve.png mask. I had too many labels due to noise in the segmentation, so this helps getting the right number of fascicles.

edgark31 commented:
> Left some comments again. Overall, this is great. Apart from the nerve indexing issue and the `cleanup_nerve_mask` function, I'm pretty satisfied with everything here.
> 
> One small nitpick is that the `unit` fields in the JSON make the file less readable. We should remove these and document it differently, i.e. in the doc itself.

I removed the  `cleanup_nerve_mask` function and tested your suggestion and i get this:
``{
    "fascicle_areas": {
        "1": {
            "value": 48690.451271182494,
            "unit": "um^2",
            "axon_density": {
                "value": 0.0,
                "unit": "axon/um^2"
            }
        },
        "2": {
            "value": 10953.742360299999,
            "unit": "um^2",
            "axon_density": {
                "value": 0.0,
                "unit": "axon/um^2"
            }
        },
        "3": {
            "value": 2095.964631975,
            "unit": "um^2",
            "axon_density": {
                "value": 0.0,
                "unit": "axon/um^2"
            }
        },
        "4": {
            "value": 118573.7997701225,
            "unit": "um^2",
            "axon_density": {
                "value": 0.0,
                "unit": "axon/um^2"
            }
        },
        "5": {
            "value": 40601.369194362494,
            "unit": "um^2",
            "axon_density": {
                "value": 0.03025,
                "unit": "axon/um^2"
            }
        }
    },
    "total_area": {
        "value": 220915.32722794247,
        "unit": "um^2"
    },
    "total_axon_density": {
        "value": 0.01942,
        "unit": "axon/um^2"
    }
}

hermancollin commented:
This is 95% done. Edgar already added some nice doc for the feature. I'll make some minor changes and add tests so we can merge this.

hermancollin commented:
Tests are failing for python 3.11 (macos/ubuntu) and python 3.12 (ubuntu). Here is the error:

```
=================================== FAILURES ===================================
___ TestCore.test_main_cli_runs_successfully_with_valid_inputs_in_nerve_mode ___

self = <test_launch_morphometrics_computation.TestCore object at 0x146f2b090>

    @pytest.mark.integration
    def test_main_cli_runs_successfully_with_valid_inputs_in_nerve_mode(self):
        path_img = self.nerve_test_file
        expected = path_img.parent / 'image_nerve_morphometrics_expected.json'
    
        output_nerve_morphometrics = path_img.parent / 'image_nerve_morphometrics.json'
        output_nerve_index = path_img.parent / 'image_nerve_index.png'
    
        with pytest.raises(SystemExit) as pytest_wrapped_e:
            AxonDeepSeg.morphometrics.launch_morphometrics_computation.main(["-i", str(path_img), "-n", "-s", "0.00236"])
    
>       assert output_nerve_morphometrics.exists()
E       AssertionError: assert False
E        +  where False = exists()
E        +    where exists = PosixPath('/Users/runner/work/axondeepseg/axondeepseg/test/__test_files__/__test_aggregate__/all_subjects/subject 1/image_nerve_morphometrics.json').exists

test/morphometrics/test_launch_morphometrics_computation.py:351: AssertionError
----------------------------- Captured stderr call -----------------------------
usage: pytest [-h] [-s SIZEPIXEL] -i IMGPATH [IMGPATH ...] [-f FILENAME]
              [-a AXONSHAPE] [-c] [-u]
pytest: error: unrecognized arguments: -n

```

Not clear why it doesn't recognize the `-n` option we are testing.

mathieuboudreau commented:
@hermancollin I think its that the changes in this repo arent in the latest pipy release. I'll push a release candidate version from this branch to see if the tests pass with that

Stoyan-I-A commented:
Hello @hermancollin, I have a question about this PR. I wanted to test it using our test image (in `axondeepseg/test/__test_files__/__test_demo_files__`) but since we don't provide an example nerve segmentation, I just created my own quickly, this is what it looks like:

![image_seg-nerve](https://github.com/user-attachments/assets/565228bf-2d14-4897-8dcf-66432b8aecb1)

But I'm not sure if this image can be used to produce an expected behaviour. 
I tried running `axondeepseg_morphometrics -i image.png -n` but I got this error:

```
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\Stoyan\Documents\ADS_stuff\axondeepseg\ads_conda\envs\venv_ads\Scripts\axondeepseg_morphometrics.exe\__main__.py", line 7, in <module>
  File "C:\Users\Stoyan\Documents\ADS_stuff\axondeepseg\AxonDeepSeg\morphometrics\launch_morphometrics_computation.py", line 287, in main
    morph_output = get_axon_morphometrics(
                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Stoyan\Documents\ADS_stuff\axondeepseg\AxonDeepSeg\morphometrics\compute_morphometrics.py", line 149, in get_axon_morphometrics
    y0, x0 = prop_axon.centroid
    ^^^^^^
ValueError: too many values to unpack (expected 2)
```

Should I be using a more "appropriate" nerve segmentation for this image or is this a bug? And if it's the former, what should a nerve segmentation look like?

hermancollin commented:
Good catch thank you @mathieuboudreau and @Stoyan-I-A. Now this part is fixed.

hermancollin commented:
@mathieuboudreau about your suggestion - I added 6 unit tests for the new functions in _compute_morphometrics.py_. I also added an integrity test for the morphometrics CLI in nerve mode in case the axon mask is missing. Tell me what you think!
